### PR TITLE
split core library from C API

### PR DESCRIPTION
**Pull request by mgovers**
* Move all core PGM functionality to `power_grid_model_c/power_grid_model/`
  * headers can be included using `#include <power_grid_model/*.hpp>`
* Move C-API to `power_grid_model_c/power_grid_model_c`
  * header can be included using `#include "power_grid_model_c.h"`
* Split C-API tests (actual integration tests) and C++ unit tests
* Move library code to respective library-specific `CMakeLists.txt`
* Link to `Boost::boost` instead of using `include_directories(${Boost_INCLUDE_DIRS})`
* Link `Threads::Threads` if found

### CODE REVIEW

File: .readthedocs.yaml [hunk 1]
```diff
@@ -17,8 +17,8 @@ build:
       # Build package with doc requirements from pyproject.optional-dependencies
       - pip install --verbose --editable .[doc]
       # remove API DLL define in header
-      - sed -i -r "s/#define PGM.*//g" include/power_grid_model_c.h
-      - sed -i -r "s/PGM_API //g" include/power_grid_model_c.h
+      - sed -i -r "s/#define PGM.*//g" power_grid_model_c/power_grid_model_c/include/power_grid_model_c.h
+      - sed -i -r "s/PGM_API //g" power_grid_model_c/power_grid_model_c/include/power_grid_model_c.h
       # build doxygen for C header
```
Comment (mgovers) [inline .readthedocs.yaml:22]: search replace for `gh_link_head_blob`

**Comment by TonyXiang8787** [comment]
@mgovers  you also need to move the folder `templates` here to the new folder of the header file 

https://github.com/PowerGridModel/power-grid-model/tree/feature/separate-c-cpp-libraries/code_generation

**Comment by TonyXiang8787** [comment]
@mgovers  there are still many links in this documentation

https://github.com/PowerGridModel/power-grid-model/blob/feature/separate-c-cpp-libraries/docs/contribution/folder-structure.md

please fix them with new links

**Comment by mgovers** [comment]
@TonyXiang8787 i believe i went over all links. the C API reference also works as intended

**Review by TonyXiang8787 — APPROVED** [review]

**Pull request merged**